### ISSUE TITLE
ONNX Runtime 推論バックエンドの実装

### ISSUE THREAD

**Issue by kurorosu** [kurorosu/pochidetection#52]
## Summary

ONNX Runtime を使用した推論バックエンド (`OnnxBackend`) を実装し, 既存の `infer` コマンドから ONNX モデル (`.onnx`) で推論できるようにする.

## Background

現在, pochidetection は PyTorch モデルによる推論 (`PyTorchBackend`) のみをサポートしている. 一方, ONNX エクスポート機能 (`OnnxExporter`) は既に実装済みであり, `onnxruntime-gpu` も依存に含まれている.

ONNX Runtime による推論を追加することで以下のメリットが得られる:

- **デプロイの柔軟性**: PyTorch 不要の軽量な推論環境を構築可能
- **既存資産の活用**: エクスポート済み `.onnx` ファイルをそのまま推論に利用可能
- **設計整合性**: 既存 `IInferenceBackend` (Strategy) に沿った自然な拡張

### 1. `OnnxBackend` クラスの実装

`pochidetection/inference/onnx_backend.py` に `IInferenceBackend` 実装を追加する.

- `onnxruntime.InferenceSession` を内部に保持
- `infer()` は `dict[str, torch.Tensor]` を受け取り, 内部で numpy 変換して `session.run()` を実行
- 出力は `torch.Tensor` に戻して `(pred_logits, pred_boxes)` を返す
- `synchronize()` は ONNX Runtime では `pass`
- Execution Providers は `CUDAExecutionProvider` -> `CPUExecutionProvider` の順でフォールバック
- サポート対象は ONNX `tensor(float)` (fp32) のみとし, 非対応 dtype は明示エラー

### 2. `DetectionPipeline` との互換維持

`DetectionPipeline` 側の変更は最小化し, 既存の入力契約を維持する.

- `OnnxBackend` 側で numpy 変換を吸収し, パイプラインは `dict[str, torch.Tensor]` のまま動作
- `torch.no_grad()` は既存実装のまま維持 (ONNX 経路でも無害)

### 3. CLI の自動判定

`--onnx` オプションは追加せず, `-m/--model-dir` のパス拡張子でバックエンドを自動判定する.

- `*.onnx` の場合: `OnnxBackend` を使用
- それ以外 (従来どおり `save_pretrained` ディレクトリ): `PyTorchBackend` を使用

```bash
# ONNX 推論 (自動判定)
uv run pochidet-rtdetr infer -d images/ -m model.onnx

# 従来の PyTorch 推論 (自動判定)
uv run pochidet-rtdetr infer -d images/ -m work_dirs/20260124_001/best
```

### 4. `infer.py` の ONNX 対応

`pochidetection/scripts/rtdetr/infer.py` に ONNX 経路を追加する.

- `model_path.suffix == ".onnx"` で `OnnxBackend` を構築
- ONNX 時の `InferenceSaver` 出力先は `model_path.parent` を基準にする (例: `<parent>/inference_001`)
- `RTDetrImageProcessor` の読み込み規約:
  - 優先: `model_path.parent` (processor ファイルがある場合)
  - フォールバック: `config["model_name"]`
  - どちらも解決不能なら明示エラー
- `BenchmarkResult.model_path` には ONNX ファイルパスを記録
- ONNX 時の `BenchmarkResult.precision` は固定で `"fp32"` とする

### 5. テスト

- `OnnxBackend` ユニットテスト (セッション初期化, providers, infer, synchronize, dtype エラー)
- `infer.py` のバックエンド自動判定テスト (`.onnx` / 非 `.onnx`)
- ONNX 時の保存先ディレクトリ規約テスト (`model_path.parent/inference_XXX`)
- ONNX 時の `benchmark_result.json` に `precision == "fp32"` が出力されることのテスト
- 出力同等性テスト: PyTorch と ONNX の `pred_logits` / `pred_boxes` を `np.allclose` で比較

## Out of Scope

- ONNX モデルの量子化 (INT8 等)
- 動的バッチサイズ対応 (現在はバッチサイズ 1 固定)
- TensorRT Execution Provider の対応
- `Detector` クラスへの ONNX 対応 (今回は `infer` スクリプト経由のみ)
- ONNX fp16 推論対応

## Acceptance Criteria

- `OnnxBackend` が `IInferenceBackend` を実装し, `.onnx` モデルで推論できる
- `uv run pochidet-rtdetr infer -d images/ -m model.onnx` で ONNX 推論が実行できる (`--onnx` 不要)
- `-m` が `.onnx` の場合は ONNX, それ以外は PyTorch が自動選択される
-  ONNX 推論時の出力先が `model.onnx` の親ディレクトリ配下 `inference_XXX/` になる
-  ONNX 推論時の `benchmark_result.json` で `precision` が `"fp32"` になる
-  PyTorch と ONNX の出力同等性を `np.allclose(rtol=1e-3, atol=1e-5)` で満たす
-  テストが追加され, `uv run pytest` が全件通る
-  `uv run pre-commit run --all-files` が通る

### 関連ファイル

| ファイル | 役割 |
|----------|------|
| `pochidetection/interfaces/backend.py` | `IInferenceBackend` インターフェース |
| `pochidetection/inference/pytorch_backend.py` | 既存の PyTorch バックエンド |
| `pochidetection/onnx/exporter.py` | ONNX エクスポート/検証ロジック |
| `pochidetection/scripts/rtdetr/infer.py` | 推論スクリプト (拡張対象) |
| `pochidetection/scripts/rtdetr/inference/detection_pipeline.py` | 推論パイプライン |
| `pochidetection/scripts/rtdetr/inference/saver.py` | 推論結果保存先の連番ディレクトリ生成 |
| `pochidetection/cli/rtdetr.py` | CLI 定義 (`infer` 引数) |

### 設計方針

- `DetectionPipeline` の入力/出力契約は維持し, バックエンド差分は `OnnxBackend` 内で吸収する
- バックエンド選択は引数追加ではなくモデルパス種別から自動判定する
- ONNX 時の `precision` は仕様として `"fp32"` 固定にする

## Labels

`enhancement`

## Branch

`feat/onnx-inference`
